here-naksha-storage-http/src/main/java/com/here/naksha/storage/http/RequestSenderCache.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (C) 2017-2023 HERE Europe B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+package com.here.naksha.storage.http;
+
+import static com.here.naksha.storage.http.RequestSender.KeyProperties;
+
+import java.util.concurrent.*;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.annotations.TestOnly;
+
+public class RequestSenderCache {
+
+  public static final int CLEANER_PERIOD_HOURS = 8;
+  private final ConcurrentMap<String, RequestSender> requestSenders;
+  private final ScheduledFuture<?> cleaner;
+
+  private RequestSenderCache() {
+    this(new ConcurrentHashMap<>(), CLEANER_PERIOD_HOURS, TimeUnit.HOURS);
+  }
+
+  /**
+   * Only for unit tests.
+   */
+  @TestOnly
```

Comment:
I don't like the fact that we had to modify production code for the sake of tests - the `@TestOnly` annotation does not really block anyone from using the annotated constructor (similarly, nothing will stop you from passing `null` to `@NotNull` argument etc).

That being said, I totally understand why you had to do that considering our discussions:
- we didn't want to use totally static class because of testability
- we couldn't use "regular" class with constructor because all of the `HttpStorage` instances need to share single `RequestSenderCache` (and we can't pass it as argument because every Storage in Naksha is constructed dynamically)
- at the same time we had touse something that has constructor that our tests can hijack and inject own props for testing 

What can we do:
- leave it as it is (I don't like it but I won't block this)
- try to remove `@TestOnly` without breaking things and still having our tests

Reply:
One thing I have in mind that keeps this as singleton but does not allow anyone from the outside to call our parameterized constructor is about:
- removing `@TestOnly`
- adding `protected` constructor modifier
- creating new private static class in your `RequestSenderCacheTest`:

```
  class TestRequestSenderCache extends RequestSenderCache {

    TestRequestSenderCache(ConcurrentMap<String, RequestSender> requestSenders, int cleanPeriod, TimeUnit cleanPeriodUnit) {
      super(requestSenders, cleanPeriod, cleanPeriodUnit);
    }
  }

```

then you run all your tests against this new class instance - everything will work as before but you don't expose constructor.

Reply:
Downside of this: we still modify production code for tests, but it's IMO a bit tighter.
We can discuss it